File: snap/snapdir/snapdir_test.go
// -*- Mode: Go; indent-tabs-mode: t -*-

/*
 * Copyright (C) 2014-2015 Canonical Ltd
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 3 as
 * published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

package snapdir_test

import (
	"errors"
	"fmt"
	"math/rand"
	"os"
	"path/filepath"
	"testing"
	"time"

	. "gopkg.in/check.v1"

	"github.com/ddkwork/golibrary/mylog"
	"github.com/snapcore/snapd/osutil"
	"github.com/snapcore/snapd/snap"
	"github.com/snapcore/snapd/snap/snapdir"
)

// Hook up check.v1 into the "go test" runner
func Test(t *testing.T) { TestingT(t) }

type SnapdirTestSuite struct{}

var _ = Suite(&SnapdirTestSuite{})

func (s *SnapdirTestSuite) TestIsSnapDir(c *C) {
	d := c.MkDir()
	mylog.Check(os.MkdirAll(filepath.Join(d, "meta"), 0755))

	mylog.Check(os.WriteFile(filepath.Join(d, "meta/snap.yaml"), nil, 0644))


	c.Check(snapdir.IsSnapDir(d), Equals, true)
}

func (s *SnapdirTestSuite) TestNotIsSnapDir(c *C) {
	c.Check(snapdir.IsSnapDir("/not-existent"), Equals, false)
	c.Check(snapdir.IsSnapDir("/dev/null"), Equals, false)
	c.Check(snapdir.IsSnapDir(c.MkDir()), Equals, false)
}

func (s *SnapdirTestSuite) TestReadFile(c *C) {
	d := c.MkDir()
	needle := []byte(`stuff`)
	mylog.Check(os.WriteFile(filepath.Join(d, "foo"), needle, 0644))


	sn := snapdir.New(d)
	content := mylog.Check2(sn.ReadFile("foo"))

	c.Assert(content, DeepEquals, needle)
}

func (s *SnapdirTestSuite) TestReadlink(c *C) {
	d := c.MkDir()
	c.Assert(os.Symlink("target", filepath.Join(d, "foo")), IsNil)

	sn := snapdir.New(d)
	target := mylog.Check2(sn.ReadLink("foo"))

	c.Assert(target, DeepEquals, "target")
}

func (s *SnapdirTestSuite) TestLstat(c *C) {
	d := c.MkDir()
	c.Assert(os.Symlink("target", filepath.Join(d, "symlink")), IsNil)
	c.Assert(os.MkdirAll(filepath.Join(d, "meta"), 0755), IsNil)
	c.Assert(os.WriteFile(filepath.Join(d, "meta/snap.yaml"), nil, 0644), IsNil)

	sn := snapdir.New(d)
	for _, file := range []string{
		"symlink",
		"meta",
		"meta/snap.yaml",
	} {
		expectedInfo := mylog.Check2(os.Lstat(filepath.Join(d, file)))

		info := mylog.Check2(sn.Lstat(file))


		c.Check(info.Name(), Equals, expectedInfo.Name())
		c.Check(info.Mode(), Equals, expectedInfo.Mode())
		c.Check(info.Size(), Equals, expectedInfo.Size())
	}
}

func (s *SnapdirTestSuite) TestLstatErrNotExist(c *C) {
	sn := snapdir.New(c.MkDir())
	_ := mylog.Check2(sn.Lstat("meta/non-existent"))
	c.Check(errors.Is(err, os.ErrNotExist), Equals, true)
}

func (s *SnapdirTestSuite) TestRandomAccessFile(c *C) {
	d := c.MkDir()
	needle := []byte(`stuff`)
	mylog.Check(os.WriteFile(filepath.Join(d, "foo"), needle, 0644))


	sn := snapdir.New(d)
	r := mylog.Check2(sn.RandomAccessFile("foo"))

	defer r.Close()

	c.Assert(r.Size(), Equals, int64(5))

	b := make([]byte, 2)
	n := mylog.Check2(r.ReadAt(b, 2))

	c.Assert(n, Equals, 2)
	c.Check(string(b), Equals, "uf")
}

func (s *SnapdirTestSuite) TestListDir(c *C) {
	d := c.MkDir()
	mylog.Check(os.MkdirAll(filepath.Join(d, "test"), 0755))

	mylog.Check(os.WriteFile(filepath.Join(d, "test", "test1"), nil, 0644))

	mylog.Check(os.WriteFile(filepath.Join(d, "test", "test2"), nil, 0644))


	sn := snapdir.New(d)
	fileNames := mylog.Check2(sn.ListDir("test"))

	c.Assert(fileNames, HasLen, 2)
	c.Check(fileNames[0], Equals, "test1")
	c.Check(fileNames[1], Equals, "test2")
}

func (s *SnapdirTestSuite) TestInstall(c *C) {
	tryBaseDir := c.MkDir()
	sn := snapdir.New(tryBaseDir)

	varLibSnapd := c.MkDir()
	targetPath := filepath.Join(varLibSnapd, "foo_1.0.snap")
	didNothing := mylog.Check2(sn.Install(targetPath, "unused-mount-dir", nil))

	c.Check(didNothing, Equals, false)
	symlinkTarget := mylog.Check2(filepath.EvalSymlinks(targetPath))

	c.Assert(symlinkTarget, Equals, tryBaseDir)
}

func (s *SnapdirTestSuite) TestInstallMustNotCrossDevices(c *C) {
	tryBaseDir := c.MkDir()
	sn := snapdir.New(tryBaseDir)

	varLibSnapd := c.MkDir()
	targetPath := filepath.Join(varLibSnapd, "foo_1.0.snap")
	didNothing := mylog.Check2(sn.Install(targetPath, "unused-mount-dir", &snap.InstallOptions{MustNotCrossDevices: true}))

	c.Check(didNothing, Equals, false)
	// TODO:UC20: fix this test when snapdir Install() understands/does
	//            something with opts.MustNotCrossDevices
	c.Check(osutil.IsSymlink(targetPath), Equals, true)
}

func walkEqual(tryBaseDir, sub string, c *C) {
	fpw := map[string]os.FileInfo{}
	filepath.Walk(filepath.Join(tryBaseDir, sub), func(path string, info os.FileInfo, err error) error {
		path = mylog.Check2(filepath.Rel(tryBaseDir, path))

		fpw[path] = info
		return nil
	})

	sdw := map[string]os.FileInfo{}
	sn := snapdir.New(tryBaseDir)
	sn.Walk(sub, func(path string, info os.FileInfo, err error) error {
		sdw[path] = info
		return nil
	})

	for k, v := range fpw {
		c.Check(os.SameFile(sdw[k], v), Equals, true, Commentf(k))
	}

	for k, v := range sdw {
		c.Check(os.SameFile(fpw[k], v), Equals, true, Commentf(k))
	}
}

func (s *SnapdirTestSuite) TestWalk(c *C) {
	// probably already done elsewhere, but just in case
	rand.Seed(time.Now().UTC().UnixNano())

	// https://en.wikipedia.org/wiki/Metasyntactic_variable
	ns := []string{
		"foobar", "foo", "bar", "baz", "qux", "quux", "quuz", "corge",
		"grault", "garply", "waldo", "fred", "plugh", "xyzzy", "thud",
		"wibble", "wobble", "wubble", "flob", "blep", "blah", "boop",
	}
	p := 1.0 / float32(len(ns))

	var f func(string, int)
	f = func(d string, n int) {
		for _, b := range ns {
			d1 := filepath.Join(d, fmt.Sprintf("%s%d", b, n))
			c.Assert(os.Mkdir(d1, 0755), IsNil)
			if n < 20 && rand.Float32() < p {
				f(d1, n+1)
			}
		}
	}

	subs := make([]string, len(ns)+3)
	// three ways of saying the same thing ¯\_(ツ)_/¯
	copy(subs, []string{"/", ".", ""})
	copy(subs[3:], ns)
	for i := 0; i < 10; i++ {
		tryBaseDir := c.MkDir()
		f(tryBaseDir, 1)

		for _, sub := range subs {
			walkEqual(tryBaseDir, sub, c)
		}

	}
}

File: cmd/libsnap-confine-private/snap-dir-test.c
/*
 * Copyright (C) 2024 Canonical Ltd
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 3 as
 * published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

#include "snap-dir.h"
#include "snap-dir.c"

#include <glib.h>
#include <stdint.h>

#include "test-utils.h"  // For rm_rf_tmp

// For compatibility with g_test_queue_destroy.
static void close_noerr(uintptr_t fd) { (void)close((int)fd); }

static void test_sc_probe_snap_mount_dir__absent(void) {
    char *root_dir = g_dir_make_tmp(NULL, NULL);
    g_assert_nonnull(root_dir);
    g_test_queue_free(root_dir);
    g_test_queue_destroy((GDestroyNotify)rm_rf_tmp, root_dir);

    int root_fd = open(root_dir, O_PATH | O_DIRECTORY | O_NOFOLLOW | O_CLOEXEC);
    g_assert_cmpint(root_fd, !=, -1);
    g_test_queue_destroy((GDestroyNotify)close_noerr, (void *)(uintptr_t)root_fd);

    sc_probe_snap_mount_dir_from_pid_1_mount_ns(root_fd, NULL);
    const char *snap_mount_dir = sc_snap_mount_dir(NULL);
    g_assert_nonnull(snap_mount_dir);
    g_assert_cmpstr(snap_mount_dir, ==, "/var/lib/snapd/snap");
}

static void test_sc_probe_snap_mount_dir__canonical(void) {
    char *root_dir = g_dir_make_tmp(NULL, NULL);
    g_assert_nonnull(root_dir);
    g_test_queue_free(root_dir);
    g_test_queue_destroy((GDestroyNotify)rm_rf_tmp, root_dir);

    char *proc_snap_dir = g_build_filename(root_dir, "/proc/1/root/snap", NULL);
    g_test_queue_free(proc_snap_dir);
    int ret = g_mkdir_with_parents(proc_snap_dir, 0755);
    g_assert_cmpint(ret, ==, 0);

    int root_fd = open(root_dir, O_PATH | O_DIRECTORY | O_NOFOLLOW | O_CLOEXEC);
    g_assert_cmpint(root_fd, !=, -1);
    g_test_queue_destroy((GDestroyNotify)close_noerr, (void *)(uintptr_t)root_fd);

    sc_probe_snap_mount_dir_from_pid_1_mount_ns(root_fd, NULL);
    const char *snap_mount_dir = sc_snap_mount_dir(NULL);
    g_assert_nonnull(snap_mount_dir);
    g_assert_cmpstr(snap_mount_dir, ==, "/snap");
}

static void test_sc_probe_snap_mount_dir__alternate_absolute(void) {
    char *root_dir = g_dir_make_tmp(NULL, NULL);
    g_assert_nonnull(root_dir);
    g_test_queue_free(root_dir);
    g_test_queue_destroy((GDestroyNotify)rm_rf_tmp, root_dir);

    char *proc_root_dir = g_build_filename(root_dir, "/proc/1/root/", NULL);
    g_test_queue_free(proc_root_dir);
    int ret = g_mkdir_with_parents(proc_root_dir, 0755);
    g_assert_cmpint(ret, ==, 0);

    char *proc_snap_symlink = g_build_filename(proc_root_dir, "snap", NULL);
    g_test_queue_free(proc_snap_symlink);
    ret = symlink("/var/lib/snapd/snap", proc_snap_symlink);
    g_assert_cmpint(ret, ==, 0);

    int root_fd = open(root_dir, O_PATH | O_DIRECTORY | O_NOFOLLOW | O_CLOEXEC);
    g_assert_cmpint(root_fd, !=, -1);
    g_test_queue_destroy((GDestroyNotify)close_noerr, (void *)(uintptr_t)root_fd);

    sc_probe_snap_mount_dir_from_pid_1_mount_ns(root_fd, NULL);
    const char *snap_mount_dir = sc_snap_mount_dir(NULL);
    g_assert_nonnull(snap_mount_dir);
    g_assert_cmpstr(snap_mount_dir, ==, "/var/lib/snapd/snap");
}

static void test_sc_probe_snap_mount_dir__alternate_relative(void) {
    char *root_dir = g_dir_make_tmp(NULL, NULL);
    g_assert_nonnull(root_dir);
    g_test_queue_free(root_dir);
    g_test_queue_destroy((GDestroyNotify)rm_rf_tmp, root_dir);

    char *proc_root_dir = g_build_filename(root_dir, "/proc/1/root/", NULL);
    g_test_queue_free(proc_root_dir);
    int ret = g_mkdir_with_parents(proc_root_dir, 0755);
    g_assert_cmpint(ret, ==, 0);

    char *proc_snap_symlink = g_build_filename(proc_root_dir, "snap", NULL);
    g_test_queue_free(proc_snap_symlink);
    ret = symlink("var/lib/snapd/snap", proc_snap_symlink);
    g_assert_cmpint(ret, ==, 0);

    int root_fd = open(root_dir, O_PATH | O_DIRECTORY | O_NOFOLLOW | O_CLOEXEC);
    g_assert_cmpint(root_fd, !=, -1);
    g_test_queue_destroy((GDestroyNotify)close_noerr, (void *)(uintptr_t)root_fd);

    sc_probe_snap_mount_dir_from_pid_1_mount_ns(root_fd, NULL);
    const char *snap_mount_dir = sc_snap_mount_dir(NULL);
    g_assert_nonnull(snap_mount_dir);
    g_assert_cmpstr(snap_mount_dir, ==, "/var/lib/snapd/snap");
}

static void test_sc_probe_snap_mount_dir__bad_symlink_target(void) {
    char *root_dir = g_dir_make_tmp(NULL, NULL);
    g_assert_nonnull(root_dir);
    g_test_queue_free(root_dir);
    g_test_queue_destroy((GDestroyNotify)rm_rf_tmp, root_dir);

    char *proc_root_dir = g_build_filename(root_dir, "/proc/1/root/", NULL);
    g_test_queue_free(proc_root_dir);
    int ret = g_mkdir_with_parents(proc_root_dir, 0755);
    g_assert_cmpint(ret, ==, 0);

    char *proc_snap_symlink = g_build_filename(proc_root_dir, "snap", NULL);
    g_test_queue_free(proc_snap_symlink);
    ret = symlink("/potato", proc_snap_symlink);
    g_assert_cmpint(ret, ==, 0);

    int root_fd = open(root_dir, O_PATH | O_DIRECTORY | O_NOFOLLOW | O_CLOEXEC);
    g_assert_cmpint(root_fd, !=, -1);
    g_test_queue_destroy((GDestroyNotify)close_noerr, (void *)(uintptr_t)root_fd);

    struct sc_error *err = NULL;
    (void)sc_probe_snap_mount_dir_from_pid_1_mount_ns(root_fd, &err);
    g_test_queue_destroy((GDestroyNotify)sc_error_free, err);
    g_assert_cmpstr(sc_error_domain(err), ==, SC_SNAP_DOMAIN);
    g_assert_cmpint(sc_error_code(err), ==, SC_SNAP_MOUNT_DIR_UNSUPPORTED);
    g_assert_cmpstr(sc_error_msg(err), ==, "/snap must be a symbolic link to /var/lib/snapd/snap");
}

static void test_sc_snap_mount_dir__not_probed(void) {
    _snap_mount_dir = NULL;

    struct sc_error *err = NULL;
    const char *snap_mount_dir = sc_snap_mount_dir(&err);
    g_test_queue_destroy((GDestroyNotify)sc_error_free, err);
    g_assert_null(snap_mount_dir);
    g_assert_cmpstr(sc_error_domain(err), ==, SC_LIBSNAP_DOMAIN);
    g_assert_cmpint(sc_error_code(err), ==, SC_API_MISUSE);
    g_assert_cmpstr(sc_error_msg(err), ==, "sc_probe_snap_mount_dir_from_pid_1_mount_ns was not called yet");
}

static void __attribute__((constructor)) init(void) {
    g_test_add_func("/snap-dir/probe-absent", test_sc_probe_snap_mount_dir__absent);
    g_test_add_func("/snap-dir/probe-canonical", test_sc_probe_snap_mount_dir__canonical);
    g_test_add_func("/snap-dir/probe-alternate-absolute", test_sc_probe_snap_mount_dir__alternate_absolute);
    g_test_add_func("/snap-dir/probe-alternate-relative", test_sc_probe_snap_mount_dir__alternate_relative);
    g_test_add_func("/snap-dir/probe-bad-symlink-target", test_sc_probe_snap_mount_dir__bad_symlink_target);
    g_test_add_func("/snap-dir/dir-not-probed", test_sc_snap_mount_dir__not_probed);
}
